Add tests to verify the safety of reportlab updates

joxl opened the pull request:
## Product Description

Updates the invoice PDF rendering logic.  Changes are expected to have no new effect on the invoice appearance.

Code review: commit by commit :tropical_fish: 

## Technical Summary

Check rendered sample invoices into version control and add tests to verify that rendering new invoices results in identical PDF content.  Created to verify #31331 (hint: the update is safe).  That PR is now on hold until this one gets merged.

## Safety Assurance

### Safety story

PR is mostly testing code. One small change was made to the invoice rendering logic, which has been verified that it looks correct on an invoice PDF.

### Automated test coverage

New tests.

### QA Plan

QA ticket: [QA-4167](https://dimagi-dev.atlassian.net/browse/QA-4167)

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

joxl committed:
Add tests to verify PDF contents rendered with ReportLab

Use `ImageReader` for invoice test logo (rather than file path) so tests pass
on systems where the logo's absolute path differs.

joxl committed:
Improve invoice PDF logo logic and add comments for the next person

- improve logo draw params to correctly position logos of any size
  (previous params only worked for a logo sized exactly 2550x1200, any
  other size would erratically move to other areas on the doc)
- add some TODO comments on known problems or unclear scenarios

millerdev reviewed (approved):
Thanks for doing this!

Edit: looks like the failing test is relevant.

gherceg commented on corehq/apps/accounting/tests/test_invoice_pdf.py:
```diff
@@ -0,0 +1,171 @@
+
+    def get_invoice(self, file_or_path, *, is_wire, is_customer, is_prepayment):
+        """Returns an invoice instantiated with all available args passed by
+        position (positionally?).
```
nit: pick one or the other? Both "passed positionally" or "passed by position" seem fine.

joxl replied:
I like "positionally", I'm going to change it.

gherceg reviewed (approved):
Just nits. Thanks @joxl!

joxl commented on corehq/apps/accounting/tests/test_invoice_pdf.py:
```diff
@@ -0,0 +1,171 @@
+            country="Around the World",
+            phone_number="",
+            email=f"{name.lower()}@example.com",
+            website="",  # populating this for a bank address messes up the footer
```
Yeah, this method is used to create `Address` objects that the test uses for all of `From`, `To`, and `Bank` addresses. The report is built in such a way that if you add a website to a bank address, it will likely cause the bank footer to wrap, which garbles the doc. I'll improve this comment.

joxl committed:
Add new `artifact` utility for writing file artifacts on test failures

- include artifacts directory in `.gitignore`
- new github action step to upload artifacts

joxl committed:
Change `InvoiceTemplate` to not assume all logos are relative file paths

- might be an absolute path, or an `ImageReader` object

joxl commented:
Tests now passing.

millerdev commented on .github/workflows/tests.yml:
```diff
@@ -51,3 +51,11 @@ jobs:
       if: always()
       run: scripts/docker down
+    - name: Upload test artifacts
+      if: always()
```
Why is this `if` clause necessary? Seems like no.

joxl replied:
Maybe I'm missing something about how GH Actions `steps` work. Are you saying this step will always run (perhaps the `strategy.fail-fast` ensures this, or is it that the previous job will never fail)?

millerdev reviewed (approved):
What do you think of removing the png logo experiment commit(s) from this PR (via rebase) to reduce churn in binary assets (files) that will be forever encoded in git history?

joxl commented:
> What do you think of removing the png logo experiment commit(s) from this PR (via rebase) ...

I'm all for this! I didn't want to do it without reviewer approval. I'll do this now.

joxl commented:
Refactored commit history to remove extra binaries and force-pushed.

joxl commented on corehq/apps/accounting/tests/test_invoice_pdf.py:
```diff
@@ -0,0 +1,188 @@
+            rendered.seek(0)
+            # write the rendered PDF as an artifact if the test fails
+            with (
+                open(fpath, "rb") as expected_pdf,
```
@stickler-ci says this is a syntax error. It's not.
